tests/end_to_end.rs

```diff
@@ -1616,3 +1620,120 @@ async fn test_update_statement_errors() {
+        // we want to be able to verify the disambiguate the versions below
+        sleep(Duration::from_secs(1)).await;
+        version_results.insert(version_id, results);
+        version_timestamps.insert(version_id, Utc::now().timestamp() as Timestamp);
+        sleep(Duration::from_secs(1)).await;
+    }
+
+    // Creates table with table_versions 1 (empty) and 2
+    create_table_and_insert(&context, "test_table").await;
+    record_result_and_timestamp(
+        &context,
+        2 as TableVersionId,
+        &mut version_results,
+        &mut version_timestamps,
+    )
+    .await;
+
+    // Add another partition for table_version 3
+    let plan = context
+        .plan_query("INSERT INTO test_table (some_value) VALUES (45), (46), (47)")
+        .await
+        .unwrap();
+    context.collect(plan).await.unwrap();
+    record_result_and_timestamp(
+        &context,
+        3 as TableVersionId,
+        &mut version_results,
+        &mut version_timestamps,
+    )
+    .await;
+
+    // Add another partition for table_version 4
+    let plan = context
+        .plan_query("INSERT INTO test_table (some_value) VALUES (46), (47), (48)")
+        .await
+        .unwrap();
+    context.collect(plan).await.unwrap();
+    record_result_and_timestamp(
+        &context,
+        4 as TableVersionId,
+        &mut version_results,
+        &mut version_timestamps,
+    )
+    .await;
+
+    // Add another partition for table_version 5
+    let plan = context
+        .plan_query("INSERT INTO test_table (some_value) VALUES (42), (41), (40)")
+        .await
+        .unwrap();
+    context.collect(plan).await.unwrap();
+
+    //
+    // Now use the recorded timestamps to query specific earlier table versions and compare them to
+    // the recorded results for that version.
+    //
+
+    async fn query_table_version(
+        context: &DefaultSeafowlContext,
+        version_id: TableVersionId,
+        version_results: &HashMap<TableVersionId, Vec<RecordBatch>>,
+        version_timestamps: &HashMap<TableVersionId, Timestamp>,
+    ) {
+        let timestamp = Utc
+            .timestamp(version_timestamps[&version_id], 0)
+            .to_rfc3339();
+        let plan = context
+            .plan_query(format!("SELECT * FROM test_table('{}')", timestamp).as_str())
+            .await
+            .unwrap();
+        let results = context.collect(plan).await.unwrap();
+
+        assert_eq!(version_results[&version_id], results);
+    }
+
+    query_table_version(
+        &context,
+        2 as TableVersionId,
+        &version_results,
+        &version_timestamps,
+    )
+    .await;
+    query_table_version(
+        &context,
+        3 as TableVersionId,
+        &version_results,
+        &version_timestamps,
+    )
+    .await;
+    query_table_version(
+        &context,
+        4 as TableVersionId,
+        &version_results,
+        &version_timestamps,
+    )
+    .await;
```

Comment:
Can you also check `information_schema.tables/columns` here to see if the versioned tables pollute the advertised list of tables in the database?

Reply:
Yup, it's not polluted; added a test for it.

Reply:
Ah, makes sense since we reset the context before every query. Can something like this happen though? Assume we're running this in a single thread and, because of async code, we end up interleaving as follows:

- Query 1 (Q1) comes in: `SELECT * FROM some_table('2020-01-01...')`
- Query 2 (Q2) comes in: `SELECT * FROM information_schema.tables`
- Database struct prepared for Q2, context points to it: C->D2 (some_table)
- Database struct prepared for Q1, context points to it: C->D1 (some_table)
- Processing for Q1 to insert table versions: C -> D1(some_table, some_table:42)
- Q2 executes, lists all tables, since C is pointing to D1 we get `some_table, some_table:42`

Reply:
Good point, I guess that could happen. Even worse, if a query comes in after `Processing for Q1 to insert table versions: C -> D1(some_table, some_table:42)`, but before executing Q1, it will reset the map (removing `some_table:42`), and Q1 will error out during execution due to the missing table ref.

Reply:
An easy potential remedy for this (I think) would be to simply clone the inner context when we encounter a version reference, create a new query planner from that cloned inner context's state, and register the table into the cloned context:
```diff
diff --git a/src/context.rs b/src/context.rs
index 26e0ece..1fcec94 100644
--- a/src/context.rs
+++ b/src/context.rs
@@ -855,6 +855,9 @@ impl SeafowlContext for DefaultSeafowlContext {
                     version_processor.visit_query(&mut q);

                     if !version_processor.table_versions.is_empty() {
+                        let inner = self.inner.clone();
+                        let state = inner.state.read().clone();
+
                         version_processor.triage_version_ids(self.table_catalog.clone()).await?;
                         // We now have table_version_ids for each table with version specified; do another
                         // run over the query AST to rewrite the table.
@@ -871,9 +874,12 @@ impl SeafowlContext for DefaultSeafowlContext {
                             let table_ref = TableReference::from(full_name.as_str());
                             let table_provider = tables_by_version[&table_version_id.unwrap()].clone();

-                            if !self.inner.table_exist(table_ref)? {
-                                self.inner.register_table(table_ref, table_provider)?;
+                            if !inner.table_exist(table_ref)? {
+                                inner.register_table(table_ref, table_provider)?;
                             }
+
+                            let query_planner = SqlToRel::new(&state);
+                            return query_planner.sql_statement_to_plan(Statement::Query(q));
                         }
                     }
```

Reply:
Makes sense, let's go for that just to be safe.